### DISCUSSION TITLE
Alternative to useFieldArray?

### DISCUSSION THREAD

**Discussion by torian257x** [react-hook-form/react-hook-form#2595] [Q&A] [answered]
hi guys  🙏 and bill glad you still so active ♥ no homo

I prepared some code without using `useFieldArray` because it didnt feel very natural using it and I had some problem, namely reading default values from an endpoint. Now I could do that using `reset` supposedly, but reading all the caveats makes me a bit nervous.

What my current problem is with my current implementation is that the values get assigned to the wrong input when I remove / add stuff. Is that the reason `useFieldArray` exists in the first place?

e.g. I have

```jsx
const delSome = (index) => {
let tmyitems = [...myItems];
tmyitems .splice(index, 1);
setMyItems(tmyitems);
};

return <>
myitems.map((el, i) => {
 return <input key={i} name={`bippo[${i}]`} type="text" ref={register}/>
});
<button onClick={delSome(2)}>del</button>
</>
```

say I have 5 myItems

I write text into all inputs, then delete item 2

as a result, my inputs seem to be in a wrong state. That is where useFieldArray comes in, right?

or is there another way to do it?

**Comment by bluebill1049** [comment]
can you please provide a codesandbox thanks?

**Reply by torian257x** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-lsp4s?file=/src/index.js

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-usefieldarray-forked-490wu

**Reply by bluebill1049** [reply]
actually the usage is wrong...

**Reply by bluebill1049** [reply]
why not just use `useFieldArray`? what's the problem are you having?

**Comment by bluebill1049** [comment]
This is similar to what you try to do: https://codesandbox.io/s/6j1760jkjk (i am not recommending that)

**Reply by torian257x** [reply]
the reason "why not?" I started without useFieldArray

then I see "oh there might be a problem". Then I look at the docs

https://react-hook-form.com/advanced-usage#FieldArrays

I look at "custom hook" in the docs above, I see: "oh I don't understand well, why? why like that?"

then I click on register on the docs above and I see: "oh, is shorter. Similar to what I used before. This is better."

![image](https://user-images.githubusercontent.com/2072058/90297709-04a91e00-de55-11ea-8f22-d2132f2157f7.png)




then I implement my complex form without useFieldArray because why use it?

then I see: "oh, there is a problem. Maybe thats why useFieldArray?"

your example:
 https://codesandbox.io/s/6j1760jkjk

has bugs when add removing items. I know that you know. But I don't know to what extent there are problems.

I think the problem is:

javascript arrays in general

they can only have key => value 0,1,2,3

they cannot have random values, different order, or string keys

key doesnt change, because when you delete key 2, key 3 now becomes key 2, thus react gets confused or something


to solve this, instead of 0,1,2,3 as index, one should use UUIDs

so instead of delete(2)
one would call `delete('34bfvasdf342314234')`

that way react cannot confuse key 3 with key 2 etc. What do you think? are there more problems?

**Reply by bluebill1049** [reply]
Thanks for the detailed feedback. I would recommend using useFieldArray to avoid those issues that I have already discovered and resolved.

**Reply by torian257x** [reply]
thanks

I would recommend to mention why to useFieldArray is important and why the register approach is bad, just one example of the problems you know?

**Comment by torian257x** [comment, marked as answer]
basically one must use `useFieldArray ` because there are many problems if one decides to not use it. For example: deleting an item at some index number, will change array without react understanding what kind of change there was.
